### ISSUE TITLE
Allow For An Interface Without a NumPyro Dependence

### ISSUE THREAD

**Issue by DanWaxman** [BasisResearch/dynestyx#213]
There may be desires in some settings to avoid a direct NumPyro dependency, dealing with the tracing, etc.

**Comment by DanWaxman** [comment]
Thankfully, this should be rather doable due to some recent refactoring that was required for hierarchical sampling, anyways. 

## Filtering

For example, all filtering integrations now come with a pure JAX implementation; for example, in cuthbert integrations:

https://github.com/BasisResearch/dynestyx/blob/c70d726fe98e423d02bc8267ded8434f49e1f578/dynestyx/inference/integrations/cuthbert/discrete.py#L163-L177

The corresponding NumPyro code that surrounds that simply runs the `compute_cuthbert_filter` path, then records some `numpyro.deterministic` sites + `numpyro.factor` statements:

https://github.com/BasisResearch/dynestyx/blob/c70d726fe98e423d02bc8267ded8434f49e1f578/dynestyx/inference/integrations/cuthbert/discrete.py#L238-L287

Similar results hold for CD-Dynamax integrations.

## Simulators

Simulators are a bit more messy, but I think still doable. 

SDE simulator would be pretty straightforward. It mostly comprises a pure-JAX `solve_sde` call + a rather simple `numpyro`-based `emit_observations` call:

https://github.com/BasisResearch/dynestyx/blob/c70d726fe98e423d02bc8267ded8434f49e1f578/dynestyx/simulators.py#L819-L851

This `_emit_observations` is really not doing anything special `NumPyro`-wise:

https://github.com/BasisResearch/dynestyx/blob/c70d726fe98e423d02bc8267ded8434f49e1f578/dynestyx/simulators.py#L591-L630

ODE simulation is a similar story.

The more painful story will be discrete-time simulation. To be honest, this is kind of a mess right now, from trying to support multiple different features (perfect observations, hierarchical simulation, batched simulation, and normal simulation). It contains a pure-JAX scan:
https://github.com/BasisResearch/dynestyx/blob/c70d726fe98e423d02bc8267ded8434f49e1f578/dynestyx/simulators.py#L1012-L1047

But it also contains a lot of other stuff... I think this will be a major pain point that we have to deal with.

## Other Caveats

There are lots of places where we use the `numpyro` PRNG key machinery, since it's pretty useful. That said, there's no reason we can't roll out our own machinery to replace it. In particular, I would propose something like `dynestyx.prng_key()` that falls back to `numpyro.prng_key()` if `PYRO_STACK` exists and has a `seed` handler in it, and otherwise, generate a global random seed [objax-style](https://objax.readthedocs.io/en/latest/objax/random.html).

## General Design

How to use this numpyro-free pathway is an open question, to me. One option that makes sense is to consider a parallel pathway to `dynestyx.sample(...)`, e.g., `dynestyx.infer(...)` instead.

Another question is how to avoid lots of duplicate code between NumPyro and non-NumPyro code. One option is, where we have pathways that only use `numpyro.sample(...)`, we roll out a background-agnostic `dynestyx._sample_tensor(...)` or something, that checks if a numpyro inference context is active, uses `numpyro.sample` if so, and uses hand-rolled `prob.sample` statements otherwise. I guess this is a bit of PyMC route, which has dangers (that codebase is very, very complex), so we would need to be very careful about scope creep of this backend-agnostic interface.

**Comment by theorashid** [comment]
Aiming to attempt this in 2 weeks when back from holidays. Starting with a draft PR for `Filter`, ignore `Simulator` for now

**Comment by theorashid** [comment]
Sorry, work has been hectic since I've been back. I'll get to this next week.

**Comment by theorashid** [comment]
I've started on this today. I wrote a version that is 

```python
filt = Filter(filter_config=KFConfig(...), register_numpyro=False)
with filt:
    dsx.sample("f", dynamics, obs_times=t, obs_values=y)
loss = -filt.marginal_loglik # use in optax or whatever

# with numpyro:
with Filter(filter_config=KFConfig(...)):
    dsx.sample("f", dynamics, obs_times=t, obs_values=y)
```

But I think this is still too coupled because Filter still imports and knows about numpyro_sites and it does not need to. I'm going to try a version more like what I had written before which is 

```python
filt = Filter(filter_config=KFConfig(...))
with filt:
    dsx.sample("f", dynamics, obs_times=t, obs_values=y)
loss = -filt.marginal_loglik

# with numpyro:
with filt, NumpyroTrace(filt):
    dsx.sample("f", dynamics, obs_times=t, obs_values=y)
```

Which breaks a lot more things but I think it's better in terms of composability and separation. So I'll continue this next week (maybe over the long weekend because I'm just in Seattle and this is fun).

**Comment by DanWaxman** [comment]
Hey Theo, thanks so much!!

I agree that the first option doesn't break as cleanly as one would like. In particular, I think most functions now break pretty cleanly into "places where work happens" and "post-hoc NumPyro registration" -- this was necessary to implement plating, since we want to vmap over the "places where work happens" parts but not over anything NumPyro.

The second is along what I was imagining, but I also don't want to have another obligatory handler for whenever one wants to use NumPyro (which seems like a fairly common use case for us, though not the only one). Do you (or maybe @mattlevine22)  have thoughts on introducing something like `dsx.infer(...)` or `dsx.condition`for the "NumPyro-free" case, and keep `dsx.sample(...)` as the NumPyro case? It seemed like "`sample`" as a function name was confusing to people anyways.

Note that this wouldn't change your implementation very much -- `dsx.sample` could just be an alias for `with NumpyroTrace(): dsx.condition(...)`. In this case, I think a nice design pattern would be: `infer` returns a dictionary `returns` of information, and `NumpyroTrace` is a simple handler that just runs `log_numpyro_values(fwd(...))`. This also simplifies your proposal a bit, as you no longer have to wrap the filter implementation in `NumpyroTrace(...)`, i.e.,
```python
# proposal from before
with filt, NumpyroTrace(filt):
    dsx.sample("f", dynamics, obs_times=t, obs_values=y)

# after, no NumPyro at all
with filt:
    dsx.infer("f", dynamics, obs_times=t, obs_values=y) 
    # or maybe dsx.condition(...)?

# after
with filt, NumpyroTrace(): # <--- no need to pass in filt
    dsx.infer("f", dynamics, obs_times=t, obs_values=y)
# or equivalently 
with filt:
    dsx.sample("f", dynamics, obs_times=t, obs_values=y)
```
This is a bit less general (I guess in your proposal, you could attach a NumPyro trace to a filter but not a simulator), though I think the all-or-nothing approach is alright.

Let me know what you think! Happy to discuss more -- I'm plenty sure there's problems with what I propose, but I do think the PPL interface really benefits from as much simplicity as possible.

**Comment by mattlevine22** [comment]
Yeah I agree with Dan that the 2nd option is better, but would prefer the default behavior of `with Filter... dsx.sample(...)` does not change from user-perspective. I also agree that the non-Numpyro usage should probably NOT use `dsx.sample()`, as it gets overloaded this way. Instead, `dsx.infer` or `dsx.condition` seem reasonable to me. 

My vision is that `dsx.sample` actually just runs something like `foo = dsx.condition(...); dsx.add_sites(foo)` (and that might mean that we have an interpretation for `condition` and an interpretation for `add_sites`, rather than the current direct interpretation of  `sample`).  

Relatedly, in the non-numpyro case, I think it would be best to simply return an object, rather than attach things to `filt`.

This is also exciting because one can imagine using this to plug into other PPLs (assuming they support JAX), e.g. PyMC

**Comment by theorashid** [comment]
Alright, I've got something like this but I'm going to spend some more time on it and launch a draft PR.

```python
def sample(...):
    result = infer(...)
    if isinstance(result, InferResult) and result._register_numpyro_sites is not None:
        result._register_numpyro_sites(name)
    return result
```

```python
def test_infer_does_not_register_numpyro_sites():
    from numpyro.handlers import seed, trace

    obs_times, obs_values = _make_data()
    dynamics = _make_lti_dynamics(0.5)

    with trace() as tr, seed(rng_seed=jr.PRNGKey(0)):
        with Filter(filter_config=KFConfig(filter_source="cuthbert")):
            result = dsx.infer("f", dynamics, obs_times=obs_times, obs_values=obs_values)

    assert isinstance(result, InferResult)
    assert result.marginal_loglik is not None
    assert "f_marginal_loglik" not in tr
    assert "f_marginal_log_likelihood" not in tr

def test_sample_registers_numpyro_sites():
    from numpyro.handlers import seed, trace

    obs_times, obs_values = _make_data()
    dynamics = _make_lti_dynamics(0.5)

    with trace() as tr, seed(rng_seed=jr.PRNGKey(0)):
        with Filter(filter_config=KFConfig(filter_source="cuthbert")):
            dsx.sample("f", dynamics, obs_times=obs_times, obs_values=obs_values)

    assert "f_marginal_loglik" in tr
    assert jnp.isfinite(tr["f_marginal_loglik"]["value"])
```

`Smoother` I could probably add. But getting the `numpyro.deterministic` out of `Simulator` might break a few more things.

So I'll just start with `Filter` draft PR, then let's discuss the design and expand once we've settled on it.

**Comment by mattlevine22** [comment]
This looks great to me! 

One question---why is `dsx.infer` inside a numpyro trace in the test? I would not expect this to be a typical use case.

**Comment by theorashid** [comment]
Yeah I don't expect anyone to do `dsx.infer` inside trace in practice. The test was to make sure the numpyro artifacts produced by `_register_numpyro_sites` were _not_ in the trace for `infer` but are there for `sample`.
